### ISSUE TITLE
Passing opaque rust types to C++ and back

### ISSUE THREAD

**Issue by smessmer** [dtolnay/cxx#802]
I'm probably missing something but I failed to get the passing of an opaque Rust type to a C++ function and back working.

header.h:
```cpp
#include "lib.rs"

void cppfunc(const MyCustomType& v) {
  rustfunc(v);
}
```

lib.rs:
```rust
#[cxx::bridge]
mod ffi {
    unsafe extern "C++" {
        include!("header.h");
        fn cppfunc(v: &MyCustomType);
    }

    extern "Rust" {
        type MyCustomType;
        fn rustfunc(v: &MyCustomType);
    }
}
struct MyCustomType()
```

The problem with this approach is an include cycle. `header.h` needs the definition of `rustfunc()` to be present, which is in `lib.rs`. And `lib.rs` needs the definition of `cppfunc` to be present, which is in `header.h`.

I could move stuff into .cpp files, but then it just becomes a circular dependency at build time between the C++ and the Rust build module.

Another potential solution I thought about is to put the definition of MyCustomType into a separate crate, and have both `lib.rs` and `header.h` depend on that third crate. This way, the dependency cycle would be broken. But this doesn't work because cxx.rs currently requires `extern "Rust"` types to be defined in the same crate where they're used.

Am I missing something? Is there really no way to pass opaque Rust types to C++ and then back?

**Comment by dtolnay** [comment]
I made #803 with a runnable example showing what I would suggest to do.

<pre>
<i>cxx</i>$  <b>git fetch origin refs/pull/803/head</b>
<i>cxx</i>$  <b>git checkout FETCH_HEAD</b>
<i>cxx</i>$  <b>cd smessmer/</b>
<i>cxx/smessmer</i>$  <b>cargo run</b>
   Compiling smessmer v0.0.0
    Finished dev [unoptimized + debuginfo] target(s) in 2.06s
     Running `target/debug/smessmer`
success
</pre>

**Comment by dtolnay** [comment]
Or as you proposed with the two directions in separate crates -- that would work too. See that one in #804.

**Comment by smessmer** [comment]
Thanks for getting back so quickly. I think for #803, it's buck that prevents my efforts since buck enforces a strict dependency DAG without cycles (and including headers in buck count as a dependency).

The separate crate approach got me a bit further, adding the `#[derive(ExternType)]` and then making it a type alias in the `extern "C++"` section seems to be the trick, as long as the custom type is just the toy type mentioned above. But with an actual custom type, I'm now getting a different error saying that my custom type is "not FFI-safe". My custom type is an attempt at pushing closures across an ffi boundary:

```rust
#[cxx::bridge(namespace="facebook::rustvelox::test")]
mod ffi {
    extern "Rust" {
        #[derive(ExternType)]
        type Closure_i64_usize;
        fn call(self: &Closure_i64_usize, v: usize) -> i64;
    }
}

pub struct Closure_i64_usize(pub Box<dyn Fn(usize) -> i64>);
impl Closure_i64_usize {
    pub fn call(&self, v: usize) -> i64 {
        self.0(v)
    }
}
```

Is this supposed to work or am I hitting a limitation of the framework?

**Comment by dtolnay** [comment]
> it's buck that prevents my efforts since buck enforces a strict dependency DAG without cycles (and including headers in buck count as a dependency).

Buck should have no problem with this; I know of code doing this pattern successfully with Buck. I will follow up with you privately.

> Is this supposed to work or am I hitting a limitation of the framework?

The "extern block uses type Closure_i64_usize, which is not FFI-safe" was a bug in old versions fixed in cxx 1.0.37. It's likely you are building with a version older than that.

**Comment by smessmer** [comment]
thanks, the upgrade to a newer cxx version fixed it.
